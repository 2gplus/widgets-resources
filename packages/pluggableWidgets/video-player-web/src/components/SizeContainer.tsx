import { CSSProperties, Component, createElement } from "react";
import classNames from "classnames";

export type HeightUnitType = "percentageOfWidth" | "percentageOfParent" | "pixels" | "aspectRatio";

export type WidthUnitType = "percentage" | "pixels";

export interface Dimensions {
    widthUnit: WidthUnitType;
    width: number;
    heightUnit: HeightUnitType;
    height: number;
    tabIndex?: number;
}

export interface SizeProps extends Dimensions {
    className: string;
    classNameInner?: string;
    style?: CSSProperties;
}

export class SizeContainer extends Component<SizeProps> {
    render(): JSX.Element {
        return (
            <div
                className={classNames(this.props.className, "size-box")}
                tabIndex={this.props.tabIndex}
                style={this.setDimensions()}
            >
                <div className={classNames("size-box-inner", this.props.classNameInner)}>{this.props.children}</div>
            </div>
        );
    }

    private setDimensions(): CSSProperties {
        const { heightUnit, height, widthUnit, width } = this.props;

        const style: CSSProperties = {};
        style.width = widthUnit === "percentage" ? `${width}%` : `${width}px`;

        switch (heightUnit) {
            case "pixels":
                style.paddingTop = height;
                break;
            case "percentageOfWidth":
                const actualHeight = (height / 100) * width;

                if (widthUnit === "percentage") {
                    style.paddingTop = `${actualHeight}%`;
                } else {
                    style.paddingTop = actualHeight;
                }
                break;
            case "percentageOfParent":
                style.paddingTop = 0;
                style.height = `${height}%`;
                break;
            case "aspectRatio":
                if (widthUnit === "pixels") {
                    style.paddingTop = width * 0.5625;
                } else {
                    style.paddingTop = `${width * 0.5625}%`; // Default is 16:9
                }
                break;
        }

        return style;
    }
}
