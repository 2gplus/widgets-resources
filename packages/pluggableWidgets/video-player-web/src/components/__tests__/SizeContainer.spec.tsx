import { createElement, ReactNode } from "react";
import { create } from "react-test-renderer";

import { SizeContainer, SizeProps } from "../SizeContainer";

describe("Size container", () => {
    const defaultProps: SizeProps & { children: ReactNode } = {
        children: <span>Child</span>,
        className: "class-name",
        classNameInner: "class-name-inner",
        style: { backgroundColor: "aqua" },
        tabIndex: 1,
        widthUnit: "percentage",
        width: 50,
        heightUnit: "pixels",
        height: 25
    };

    it("renders correctly with height unit pixels", () => {
        const sizeContainer = create(<SizeContainer {...defaultProps} />).toJSON();

        expect(sizeContainer).toMatchSnapshot();
    });

    it("renders correctly with height unit percentage of width", () => {
        const sizeContainer = create(<SizeContainer {...defaultProps} heightUnit="percentageOfWidth" />).toJSON();

        expect(sizeContainer).toMatchSnapshot();
    });

    it("renders correctly with height unit percentage of parent", () => {
        const sizeContainer = create(<SizeContainer {...defaultProps} heightUnit="percentageOfParent" />).toJSON();

        expect(sizeContainer).toMatchSnapshot();
    });

    it("renders correctly with height unit aspect ratio", () => {
        const sizeContainer = create(
            <SizeContainer
                {...defaultProps}
                heightUnit="aspectRatio"
                height={undefined}
                heightAspectRatio="TwentyOneByNine"
            />
        ).toJSON();

        expect(sizeContainer).toMatchSnapshot();
    });
});
