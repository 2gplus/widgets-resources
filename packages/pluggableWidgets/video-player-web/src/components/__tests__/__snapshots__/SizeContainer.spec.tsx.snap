// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Size container renders correctly with height unit aspect ratio 1:1 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "50%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit aspect ratio 3:2 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "33.33333333333333%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit aspect ratio 4:3 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "37.5%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit aspect ratio 16:9 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "28.125%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit aspect ratio 21:9 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "21.428571428571427%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit percentage of parent 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "height": "25%",
      "paddingTop": 0,
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit percentage of width 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": "12.5%",
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with height unit pixels 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": 25,
      "width": "50%",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with pixel width & height unit aspect ratio 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": 21.428571428571427,
      "width": "50px",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;

exports[`Size container renders correctly with pixel width & height unit percentage of width 1`] = `
<div
  className="class-name size-box"
  style={
    Object {
      "backgroundColor": "aqua",
      "paddingTop": 12.5,
      "width": "50px",
    }
  }
  tabIndex={1}
>
  <div
    className="size-box-inner class-name-inner"
  >
    <span>
      Child
    </span>
  </div>
</div>
`;
